### PR TITLE
[cssom-view-1] Added definition of getBoxQuads()

### LINKED ISSUE #10537: [css-omview] update description of getBoxQuads API
The getBoxQuads API is missing lot of its description.

Issue in Firefox:
https://bugzilla.mozilla.org/show_bug.cgi?id=1107559

Description 
https://hacks.mozilla.org/2014/03/introducing-the-getboxquads-api/

### PR DESCRIPTION
**Pull request by jogibear9988**
Fixes #10537

A linked issue:
#10514

A firefox issue with a little bit of description:
https://bugzilla.mozilla.org/show_bug.cgi?id=1107559

A polyfill for the API
https://github.com/jogibear9988/getBoxQuadsPolyfill

### CODE REVIEW

**Comment by jogibear9988** [comment]
One issues atm:
My polyfill only works for 2D transformations, 3D with perspective do not work (yet). Have to figure out how to do this. Cause of that, also my spec only works for 2D. I've found some spect for the 3D matrix, wich then also could maybe reused: https://drafts.csswg.org/css-transforms-2/#accumulated-3d-transformation-matrix-computation

**Comment by SebastianZ** [comment]
I think it makes more sense to add another option to `getBoxQuads()` to consider the transformed coordinates instead of introducing a separate function `getCompleteTransform()` for that.

I imagine something like `getBoxQuads({considerTransformations: true})`.

Sebastian

**Comment by jogibear9988** [comment]
> One issues atm: My polyfill only works for 2D transformations, 3D with perspective do not work (yet). Have to figure out how to do this. Cause of that, also my spec only works for 2D. I've found some spect for the 3D matrix, wich then also could maybe reused: https://drafts.csswg.org/css-transforms-2/#accumulated-3d-transformation-matrix-computation

should be fixed in current polyfill

**Comment by jogibear9988** [comment]
> I think it makes more sense to add another option to `getBoxQuads()` to consider the transformed coordinates instead of introducing a separate function `getCompleteTransform()` for that.
> 
> I imagine something like `getBoxQuads({considerTransformations: true})`.
> 
> Sebastian

it's in a extra method, cause it is used from the 3 other functions in the spec

File: cssom-view-1/Overview.bs [hunk 1]
```diff
@@ -1781,46 +1781,76 @@ Document includes GeometryUtils;
 typedef (Text or Element or CSSPseudoElement or Document) GeometryNode;
 </pre>
 
-The <dfn method for=GeometryUtils lt="getBoxQuads(options)|getBoxQuads()">getBoxQuads(<var>options</var>)</dfn> method must run the following steps:
-
+The helper method getCompleteTransform(anchestor) must run the following steps:
```
Comment (emilio) [inline cssom-view-1/Overview.bs:1784]: You need to get the element as a parameter, and use `<div algorithm>` for this for example:

```
<div algorithm="get the complete transform">
  In order to <dfn export>get the complete transform</dfn> of an node <var>node</var> relative to a node <var>ancestor</ancestor> ...
</div>

File: cssom-view-1/Overview.bs [hunk 2]
```diff
@@ -1781,46 +1781,76 @@ Document includes GeometryUtils;
...
-
-  scale to 0 means divide by zero, return 0x0
-
-  cross-frames not allowed, throw WrongDocumentError?
-
-  points are flattened (3d transform), z=0. like getClientRect
-
-  test block in inline
+1. let currentElement be the element getCompleteTransform() called on.
+2. let transformationMatrix be the current transformation matrix of currentElement (https://www.w3.org/TR/css-transforms-2/#ctm)
```
Comment (emilio) [inline cssom-view-1/Overview.bs:1787]: Please use bikeshed references for this.

File: cssom-view-1/Overview.bs [hunk 3]
```diff
@@ -1781,46 +1781,76 @@ Document includes GeometryUtils;
...
+ 1. create a DOMQuad from the rect via DOMQuad.fromRect
+ 2. transform each point of the DOMQuad from 1. via convertPointFromNode
 </ol>
 
 The <dfn method for=GeometryUtils lt="convertPointFromNode(point, from, options)|convertPointFromNode(point, from)">convertPointFromNode(<var>point</var>, <var>from</var>, <var>options</var>)</dfn>
 method must run the following steps:
 
 <ol>
- <li><p class=issue>...
+1. get the resulting matrix between the element and the document.defaultView via the algorithm defined in getCompleteTransform
```
Comment (emilio) [inline cssom-view-1/Overview.bs:1838]: All these algorithms need to be in terms of `node`s, as `GeometryUtils` also applies to text nodes and such.

File: cssom-view-1/Overview.bs [hunk 4]
```diff
@@ -1781,46 +1781,76 @@ Document includes GeometryUtils;
...
+2. get the resulting matrix between the from and the document.defaultView via the algorithm defined in getCompleteTransform
+3. inverse the matrix in 2.
+4. Depending on the value of options.fromBox modify the point with the calculated style values of the element.
+   - 'border': no change
+   - 'margin': x - margin.left, y - margin.top
+   - 'padding': x + border.leftWidth, y + border.topWidth
+   - 'content': x + border.leftWidth + padding.left, y + border.topWidth + padding.top
```
Comment (emilio) [inline cssom-view-1/Overview.bs:1845]: This is not actually based on the computed style but on the actual used padding box etc. Those boxes are defined already elsewhere and we should reuse the definitions.

**Review by emilio — CHANGES_REQUESTED** [review]
Thanks, this looks like a good start. Please look at how other specs (specially recent ones) use bikeshed in order to get the formatting and references correct. Also this needs to work in terms of nodes and boxes, not elements.

**Comment by whimboo** [comment]
Hi @jogibear9988, would you maybe have the time to follow-up on the review from Emilio? Thanks

File: cssom-view-1/Overview.bs (hunk 3 above)
Comment (jogibear9988) [reply cssom-view-1/Overview.bs]: Is node correct? Or should it be "GeometryNode"?
image

File: cssom-view-1/Overview.bs (hunk 2 above)
Comment (jogibear9988) [reply cssom-view-1/Overview.bs]: Where should I use a bikeshed reference? (Is there an editor wich helps in editing bs files)

File: cssom-view-1/Overview.bs (hunk 4 above)
Comment (jogibear9988) [reply cssom-view-1/Overview.bs]: I don't really find where I should link to.
maybe: https://drafts.csswg.org/css2/#Computing_widths_and_margins

File: cssom-view-1/Overview.bs (hunk 1 above)
Comment (jogibear9988) [reply cssom-view-1/Overview.bs:1864]: I don't get what you mean with this....
getCompleteTransform has a parameter, wich is anchestor

Comment (jogibear9988) [reply cssom-view-1/Overview.bs]: Added node to helper function

File: cssom-view-1/Overview.bs (hunk 4 above)
Comment (Psychpsyo) [reply cssom-view-1/Overview.bs]: > I don't really find where I should link to. maybe: https://drafts.csswg.org/css2/#Computing_widths_and_margins

Pretty sure those boxes are defined here: https://drafts.csswg.org/css-box-4/#box-edges

Comment (jogibear9988) [reply cssom-view-1/Overview.bs]: Yeah, but I need to link to the Algorythm how they are calculated, or is this enough?

Comment (Psychpsyo) [reply cssom-view-1/Overview.bs]: > Yeah, but I need to link to the Algorythm how they are calculated, or is this enough?

There is no algorithm like that for these. They are defined in English instead. Browsers have algorithms internally that match those descriptions and can call into those when a spec says to do something with one of these boxes, so linking to their definitions is enough to unambiguously tell an implementer what needs to happen.

More modern specs like to define their concepts in terms of algorithms to derive them, because that makes implementation easier and leaves less room for creative (mis-)interpretation of the definition.

TL;DR: You want to link to a concept's definition when using it. Whether that is an algorithm or not doesn't matter.

**Comment by jogibear9988** [comment]
@emilio 
is it okay like now?

File: cssom-view-1/Overview.bs [hunk 5]
```diff
@@ -1861,46 +1861,74 @@ Document includes GeometryUtils;
...
-  p1 = top left even in RTL
-
-  scale to 0 means divide by zero, return 0x0
-
-  cross-frames not allowed, throw WrongDocumentError?
-
-  points are flattened (3d transform), z=0. like getClientRect
-
-  test block in inline
+ 1. let currentElement be node.
```
Comment (emilio) [inline cssom-view-1/Overview.bs]: currentNode, right? I think this should use `<var>current</var>` and other bikeshed markup to display properly.

File: cssom-view-1/Overview.bs [hunk 6]
```diff
@@ -1861,46 +1861,74 @@ Document includes GeometryUtils;
...
+ 2. let transformationMatrix be the current transformation matrix of currentElement (https://www.w3.org/TR/css-transforms-2/#ctm)
+ 3. while currentElement is not null do
+ <ol>
```
Comment (emilio) [inline cssom-view-1/Overview.bs]: Do you really need the `<ol>`?

File: cssom-view-1/Overview.bs [hunk 7]
```diff
@@ -1861,46 +1861,74 @@ Document includes GeometryUtils;
 typedef (Text or Element or CSSPseudoElement or Document) GeometryNode;
 </pre>
 
-The <dfn method for=GeometryUtils lt="getBoxQuads(options)|getBoxQuads()">getBoxQuads(<var>options</var>)</dfn> method must run the following steps:
-
+The helper method getCompleteTransform(node, anchestor) must run the following steps:
```
Comment (emilio) [inline cssom-view-1/Overview.bs]: I think you should probably add a `<dfn>` so that you can reference this properly elsewhere.

File: cssom-view-1/Overview.bs [hunk 8]
```diff
@@ -1861,46 +1861,74 @@ Document includes GeometryUtils;
...
+	 6. if currentElement is the anchestor or the anchestor is null return the transformationMatrix
+ </ol>
+</ol>
 
-  pseudo-elements before/after are children of the element
+The <dfn method for=GeometryUtils lt="getBoxQuads(options)|getBoxQuads()">getBoxQuads(<var>options</var>)</dfn> method must run the following steps:
 
-  viewport boxes are all the same
+<ol>
+ 1. create an array as result container
```
Comment (emilio) [inline cssom-view-1/Overview.bs]: `Let |result| be an empty [=list=] of [=DOMQuad=] objects.` etc, please look at existing bikeshed files to see how to make this work.

File: cssom-view-1/Overview.bs [hunk 9]
```diff
@@ -1861,46 +1861,74 @@ Document includes GeometryUtils;
...
+ 2. for each css-fragment of the element run the following steps
```
Comment (emilio) [inline cssom-view-1/Overview.bs]: of what element? Probably of `this`? Also fragment should link to a definition.

File: cssom-view-1/Overview.bs [hunk 10]
```diff
@@ -1861,46 +1861,74 @@ Document includes GeometryUtils;
...
+ <ol>
+    1. Create a DOMRect of the Element with the x,y value of 0 and the offsetWith and offsetHeight.
```
Comment (emilio) [inline cssom-view-1/Overview.bs]: This doesn't seem right, `offsetWidth` rounds to integers to begin with which this code shouldn't do, but also is not defined for non-elements, so this should probably do something like:

 * If `this` is a `Document`, then you need to use the layout viewport.
 * If `this` is an `Element` or `Text` node, then you need to iterate over the fragments.

File: cssom-view-1/Overview.bs [hunk 11]
```diff
@@ -1861,46 +1861,74 @@ Document includes GeometryUtils;
...
+    2. Depending on the value of options.box modify the Rect from 1. with the calculated style values of the element.
+    - 'border': no change
+    - 'margin': https://drafts.csswg.org/css-box-4/#box-edges - use the margin-box
```
Comment (emilio) [inline cssom-view-1/Overview.bs]: This should not just drop the urls like this, it should link to the definition. `ResizeObserver` has _very_ similar functionality, maybe crib from there?

File: cssom-view-1/Overview.bs [hunk 12]
```diff
@@ -1861,46 +1861,74 @@ Document includes GeometryUtils;
...
+    - 'padding': https://drafts.csswg.org/css-box-4/#box-edges - use the padding-box
+    - 'content': https://drafts.csswg.org/css-box-4/#box-edges - use the content-box
+    3. create a DomQuad from the rect via DOMQuad.fromRect
```
Comment (emilio) [inline cssom-view-1/Overview.bs]: Proper capitalization `DOMQuad` and linking please.

File: cssom-view-1/Overview.bs [hunk 13]
```diff
@@ -1861,46 +1861,74 @@ Document includes GeometryUtils;
...
+    4. Calculate the resulting transformation matrix between the element and options.realtiveTo (if relativeTo is unset use window.defaultView) via the algorithm defined in getCompleteTransform
```
Comment (emilio) [inline cssom-view-1/Overview.bs]: `window.defaultView` is not a thing, also `s/realtiveTo/relativeTo`.

This should be untransforming the rect to the layout viewport if there's no `relativeTo`.

File: cssom-view-1/Overview.bs [hunk 14]
```diff
@@ -1861,46 +1861,74 @@ Document includes GeometryUtils;
...
+ </ol>
+ 3. return the result container
 </ol>
 
 The <dfn method for=GeometryUtils lt="convertQuadFromNode(quad, from, options)|convertQuadFromNode(quad, from)">convertQuadFromNode(<var>quad</var>, <var>from</var>, <var>options</var>)</dfn> method
 must run the following steps:
 
 <ol>
- <li><p class=issue>...
+ 1. transform each point of the DOMQuad via convertPointFromNode
```
Comment (emilio) [inline cssom-view-1/Overview.bs:2385]: This should link to `convertPointFromNode`, but also probably should have more precise details like:

```
 * Let <var>p1</var> be the result of calling [=convertFromFromNode=] with <var>quad</var>.p1, <var>from</var> and <var>options</var>.
 * ...
 * Return a new [=DOMQuad=] with <var>p1</var>...
```

File: cssom-view-1/Overview.bs [hunk 15]
```diff
@@ -1861,46 +1861,74 @@ Document includes GeometryUtils;
...
+ 1. create a DOMQuad from the rect via DOMQuad.fromRect
+ 2. transform each point of the DOMQuad from 1. via convertPointFromNode
 </ol>
 
 The <dfn method for=GeometryUtils lt="convertPointFromNode(point, from, options)|convertPointFromNode(point, from)">convertPointFromNode(<var>point</var>, <var>from</var>, <var>options</var>)</dfn>
 method must run the following steps:
 
 <ol>
- <li><p class=issue>...
+ 1. get the resulting matrix between the node and the document.defaultView via the algorithm defined in getCompleteTransform
```
Comment (emilio) [inline cssom-view-1/Overview.bs]: Give these meaningful names like:

```
 * Let <var>thisTransformToViewport</var> be the result of calling [=getCompleteTransform=] with <var>this</var> and <var>this</var>' [=associated document=].
 * Let <var>fromTransformToViewport</var> be the result of ...
 * Invert <var>fromTransformToViewport</var>.
 * ...
```

Then reference those.

File: cssom-view-1/Overview.bs [hunk 16]
```diff
@@ -1861,46 +1861,74 @@ Document includes GeometryUtils;
...
+ 2. get the resulting matrix between the from and the document.defaultView via the algorithm defined in getCompleteTransform
+ 3. inverse the matrix in 2.
+ 4. Depending on the value of options.fromBox modify the point with the calculated style values of the element.
+   - 'border': no change
+   - 'margin': x - margin.left, y - margin.top
+   - 'padding': x + border.leftWidth, y + border.topWidth
+   - 'content': x + border.leftWidth + padding.left, y + border.topWidth + padding.top
```
Comment (emilio) [inline cssom-view-1/Overview.bs]: Same, refer to standard terms. It seems `ResizeObserver` refers to the CSS2 border box area, etc.

File: cssom-view-1/Overview.bs [hunk 17]
```diff
@@ -1861,46 +1861,74 @@ Document includes GeometryUtils;
...
 
 The <dfn method for=GeometryUtils lt="convertPointFromNode(point, from, options)|convertPointFromNode(point, from)">convertPointFromNode(<var>point</var>, <var>from</var>, <var>options</var>)</dfn>
 method must run the following steps:
 
 <ol>
- <li><p class=issue>...
+ 1. get the resulting matrix between the node and the document.defaultView via the algorithm defined in getCompleteTransform
+ 2. get the resulting matrix between the from and the document.defaultView via the algorithm defined in getCompleteTransform
+ 3. inverse the matrix in 2.
+ 4. Depending on the value of options.fromBox modify the point with the calculated style values of the element.
```
Comment (emilio) [inline cssom-view-1/Overview.bs]: This should use `<var>point</var>` etc.

**Review by emilio — CHANGES_REQUESTED** [review]
This still needs a bit of work to get the spec into a reasonable shape, but looks in the right direction. Thanks!

**Comment by whimboo** [comment]
Hi @jogibear9988. Would you maybe have the time to update this PR based on the review feedback from Emilio? Thanks a lot!

**Comment by jogibear9988** [comment]
Sorry, I've read them, and hopefully fix in the next days

**Comment by whimboo** [comment]
Hi @jogibear9988. I wanted to check back with you about the status of this PR. Is it still something you can continue working on? Thanks!

**Comment by jogibear9988** [comment]
@whimboo 
sorry, had not much time to work on this.
at the moment I fix my polyfill (https://github.com/jogibear9988/getBoxQuadsPolyfill) to also respect offset-.... css properties. Didn't know they exist. 
Also there is a small issue with webcomponents I try to fix.
Then I look to fix this spec.

**Comment by jogibear9988** [comment]
Also I think the spec needs an upgrade to also include a function to calculate the complete transformation matrix between two elements.
My polyfill does this, and I need this also some time in my designer, and this is not possible only with the boxQuads result.
is this something I also should add here, or should this be an extra issue/pull req?

**Comment by bramus** [comment]
> Also I think the spec needs an upgrade to also include a function to calculate the complete transformation matrix between two elements. My polyfill does this, and I need this also some time in my designer, and this is not possible only with the boxQuads result. is this something I also should add here, or should this be an extra issue/pull req?

This would need a separate PR – one that would need a CSS WG Resolution first to get the group’s consensus to adding it.

**Comment by whimboo** [comment]
Hello @jogibear9988. I want to check back with you if you may have some time that this PR could be moved forward? Or is it blocked based on the last comment? Thanks!

File: cssom-view-1/Overview.bs [hunk 18]
```diff
@@ -2192,26 +2194,107 @@ Document includes GeometryUtils;
...
+    1. Let <var>rect</var> be a {{DOMRect}} describing the <a>border edge</a> of
+        the <a spec=css-break>box fragment</a>,
+        with {{DOMRect/x}} and {{DOMRect/y}} set to 0,
+        and {{DOMRect/width}} and {{DOMRect/height}} set to the width and height
+        of the [=border box=] of the fragment.
+    1. If <var>options</var>.{{BoxQuadOptions/box}} is
+        {{CSSBoxType/"margin"}},
+        adjust <var>rect</var> to describe the <a>margin edge</a> of the fragment
+        by subtracting the used value of 'margin-left' from {{DOMRect/x}},
+        subtracting the used value of 'margin-top' from {{DOMRect/y}},
```
Comment (Psychpsyo) [inline cssom-view-1/Overview.bs]: These calculations are not how you get a box fragment's margin/padding/border box edges. Whether a box fragment has margins, borders and padding along any given edge depends on many factors beyond just its used values. (Whether that edge adjoins a fragmentainer break and whether or not the box is fragmented along that edge and the value of the `box-decoration-break` CSS property on that box and the value of `margin-trim` on any ancestors whose inner edge that edge touches...) I'd just say which edge the values should correspond to and that the DOMRect should be positioned so that the border edge is at 0,0 instead of trying te re-define the conversion between the different edges from first principles somehow.

Comment (jogibear9988) [reply cssom-view-1/Overview.bs]: Thanks, that makes sense.

I changed this in the latest update. `getBoxQuads()` now defines the fragment rect in local coordinates with the border edge at 0,0, and for the `"margin"`, `"padding"`, and `"content"` cases it refers to the fragment’s corresponding box area instead of re-deriving those edges from used values.

So this should now follow the standard box definitions rather than trying to reconstruct them from first principles.

**Comment by jogibear9988** [comment]
also added some description to my polyfill: https://github.com/jogibear9988/getBoxQuadsPolyfill/blob/main/csswg-getBoxQuads-calculation.md

**Comment by jogibear9988** [comment]
I also have a new library wich uses getBoxQuads API, to be able to create screenshots or vector exports of HTML pages: https://github.com/node-projects/layout2vector

**Comment by whimboo** [comment]
@jogibear9988 is this PR ready for another review? If yes, mind re-requesting review from Emilio? Thanks.

**Comment by jogibear9988** [comment]
Yeah, should be ready

File: cssom-view-1/Overview.bs [hunk 19]
```diff
@@ -2177,7 +2179,7 @@ The <dfn attribute for=MouseEvent>offsetY</dfn> attribute must follow these step
 enum CSSBoxType { "margin", "border", "padding", "content" };
 dictionary BoxQuadOptions {
   CSSBoxType box = "border";
-  GeometryNode relativeTo; // XXX default document (i.e. viewport)
+    GeometryNode relativeTo; // if omitted, coordinates are relative to the layout viewport
```
Comment (emilio) [inline cssom-view-1/Overview.bs]: nit: Indentation. Also the "if omitted.." should be in the spec prose

File: cssom-view-1/Overview.bs [hunk 20]
```diff
@@ -2189,7 +2191,7 @@ interface mixin GeometryUtils {
   sequence&lt;DOMQuad> getBoxQuads(optional BoxQuadOptions options = {});
   DOMQuad convertQuadFromNode(DOMQuadInit quad, GeometryNode from, optional ConvertCoordinateOptions options = {});
   DOMQuad convertRectFromNode(DOMRectReadOnly rect, GeometryNode from, optional ConvertCoordinateOptions options = {});
-  DOMPoint convertPointFromNode(DOMPointInit point, GeometryNode from, optional ConvertCoordinateOptions options = {}); // XXX z,w turns into 0
+    DOMPoint convertPointFromNode(DOMPointInit point, GeometryNode from, optional ConvertCoordinateOptions options = {}); // result is flattened to 2D
```
Comment (emilio) [inline cssom-view-1/Overview.bs]: nit: Ditto

File: cssom-view-1/Overview.bs [hunk 21]
```diff
@@ -2200,26 +2202,158 @@ Document includes GeometryUtils;
...
-  DOM order
-
-  p1 = top left even in RTL
-
-  scale to 0 means divide by zero, return 0x0
-
-  cross-frames not allowed, throw WrongDocumentError?
+The {{GeometryUtils}} methods operate on a node's local box geometry.
+The {{GeometryUtils/getBoxQuads()}} method returns one {{DOMQuad}} per relevant
+box fragment, in content order.
```
Comment (emilio) [inline cssom-view-1/Overview.bs]: This should be in flattened tree order, and should link to the definition.

File: cssom-view-1/Overview.bs (hunk 19 above)
Comment (emilio) [reply cssom-view-1/Overview.bs]: It's already, so just remove the comment.

File: cssom-view-1/Overview.bs [hunk 22]
```diff
@@ -2200,26 +2202,158 @@ Document includes GeometryUtils;
...
+<div algorithm="get the complete transform">
+    To <dfn export>get the complete transform</dfn> of a {{GeometryNode}} <var>node</var>
+    relative to a {{GeometryNode}} <var>ancestor</var>,
+    run the following steps:
+
+    1. Let <var>current</var> be <var>node</var>.
+    1. Let <var>transformationMatrix</var> be a new {{DOMMatrix}}.
+    1. While <var>current</var> is not null:
+        1. If <var>current</var> is <var>ancestor</var>, return <var>transformationMatrix</var>.
+        1. Let <var>parent</var> be null.
```
Comment (emilio) [inline cssom-view-1/Overview.bs]: Parent is not `parent` right? It's more of `ancestor`

**Review by emilio — CHANGES_REQUESTED** [review]
Thanks, this looks a lot better! Some nits and tweaks but looks good to me with those fixed.

**Review by emilio — APPROVED** [review]
I think this is in a pretty good state now. Could you squash please? Also I don't know what needs to happen to make the IPR bot happy... @svgeesus can you help?

**Comment by whimboo** [comment]
> I think this is in a pretty good state now. Could you squash please? Also I don't know what needs to happen to make the IPR bot happy... @svgeesus can you help?

As it looks like @jogibear9988 needs to [join the Cascading Style Sheets (CSS) Working Group](https://www.w3.org/groups/wg/css/join). Once done it should be fine.

**Comment by jogibear9988** [comment]
Don‘t think this is so easy
https://www.w3.org/groups/wg/css/instructions/

Somebody needs to invite me

**Comment by Psychpsyo** [comment]
> Don‘t think this is so easy https://www.w3.org/groups/wg/css/instructions/
> 
> Somebody needs to invite me

You can send an application to become an Invited Expert without explicitly being invited beforehand.

**Comment by jogibear9988** [comment]
I‘d send it

**Comment by svgeesus** [comment]
@jogibear9988 I will send you a request for an individual IPR commitment for this PR.

(There is no need for Invited Expert for this one PR)

**Comment by jogibear9988** [comment]
> @jogibear9988 I will send you a request for an individual IPR commitment for this PR.
> 
> (There is no need for Invited Expert for this one PR)

Okay for me

**Comment by jogibear9988** [comment]
can someone retrigger the IPR check?

**Comment by jogibear9988** [comment]
and if we have this merged, what are the next steps? Is there a way to get this going forward in browsers? (for example move the API in FF away from the flag? get it implemented in chrome? (nearly every browser already has 90% of this already for the dev tools)

**Comment by Psychpsyo** [comment]
> and if we have this merged, what are the next steps? Is there a way to get this going forward in browsers? (for example move the API in FF away from the flag? get it implemented in chrome? (nearly every browser already has 90% of this already for the dev tools)

For one, you can open browser bugs and write web platform tests for the feature (both of those should happen regardless). Implementing a well-tested feature that someone asked for is going to be an easier sell than just "spec says to do this". But the only thing that will definitely get this implemented in browsers is if you do the implementation yourself / pay someone to do it. There isn't much other special magic that can compel a browser vendor into doing something the spec tells them to, and every browser vendor has to prioritize between hundreds of things already.

**Comment by jogibear9988** [comment]
> > and if we have this merged, what are the next steps? Is there a way to get this going forward in browsers? (for example move the API in FF away from the flag? get it implemented in chrome? (nearly every browser already has 90% of this already for the dev tools)
> 
> For one, you can open browser bugs and write web platform tests for the feature (both of those should happen regardless). Implementing a well-tested feature that someone asked for is going to be an easier sell than just "spec says to do this". But the only thing that will definitely get this implemented in browsers is if you do the implementation yourself / pay someone to do it. There isn't much other special magic that can compel a browser vendor into doing something the spec tells them to, and every browser vendor has to prioritize between hundreds of things already.

there already exist some:
https://github.com/web-platform-tests/wpt/blob/master/css/cssom-view/cssom-getBoxQuads-001.html
https://github.com/web-platform-tests/wpt/blob/master/css/cssom-view/cssom-getBoxQuads-002.html

yeah, only wonder, firefox has it since 2014 (12 years), and kept it behind a flag since then.

**Comment by jogibear9988** [comment]
I added a few more tests:

https://github.com/web-platform-tests/wpt/pull/61015

**Comment by whimboo** [comment]
@jogibear9988 I also saw that you created https://github.com/web-platform-tests/wpt/pull/50314 a while ago and which is still open and not reviewed. Is that PR up-to-date so that we could try to see it reviewed as well?

Regarding the remaining steps for Firefox, we seem to still have a [couple of bugs open for `getBoxQuads`](https://bugzilla.mozilla.org/buglist.cgi?query_format=advanced&short_desc_type=allwordssubstr&short_desc=getBoxQuads&bugidtype=include&list_id=18028832), and possibly some related issues as well.

@emilio, could you maybe provide an overview of what would still be needed on our side to move this API out from behind the flag once this PR got landed?

**Comment by whimboo** [comment]
FYI I triggered the IPR revalidation and it seems to be all fine now. Now someone with permissions is needed to get it merged.

**Comment by emilio** [comment]
> @emilio, could you maybe provide an overview of what would still be needed on our side to move this API out from behind the flag once this PR got landed?

More tests would be good, but on the Gecko side the main blocker was the spec and ideally some implementation interest from other browsers, see https://bugzilla.mozilla.org/show_bug.cgi?id=2026051

**Review by emilio — APPROVED** [review]
Let's try this, thanks!

**Comment by jogibear9988** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=1107559

**Comment by whimboo** [comment]
Thanks a lot @jogibear9988 for getting it finished! It's great to see that we are closer to be able to offer support for these APIs via WebDriver BiDi (https://github.com/w3c/webdriver-bidi/issues/787).

**Comment by jogibear9988** [comment]
> > and if we have this merged, what are the next steps? Is there a way to get this going forward in browsers? (for example move the API in FF away from the flag? get it implemented in chrome? (nearly every browser already has 90% of this already for the dev tools)
> 
> For one, you can open browser bugs and write web platform tests for the feature (both of those should happen regardless). Implementing a well-tested feature that someone asked for is going to be an easier sell than just "spec says to do this". But the only thing that will definitely get this implemented in browsers is if you do the implementation yourself / pay someone to do it. There isn't much other special magic that can compel a browser vendor into doing something the spec tells them to, and every browser vendor has to prioritize between hundreds of things already.

did this:
https://chromium-review.googlesource.com/c/chromium/src/+/8039580
